url270fc4f5ee_ykcilborw_Joust_tgz-pJ8-fri_patterns_interpreter_parsergenerator_examples_HowToOverrideLexerImplJ8/src/fri/patterns/interpreter/parsergenerator/builder/SerializedObject.java

```diff
@@ -0,0 +1,129 @@
+package fri.patterns.interpreter.parsergenerator.builder;
+
+import java.io.*;
+import java.util.*;
+import fri.util.props.ConfigDir;
+import fri.patterns.interpreter.parsergenerator.syntax.*;
+import fri.patterns.interpreter.parsergenerator.syntax.builder.*;
+
+/**
+	Base class for serialization and deserialisation of Java-objects.
+	
+	@author (c) 2000, Fritz Ritzberger
+*/
+
+class SerializedObject
+{
+	/** Do-nothing constructor. */
+	public SerializedObject()	{
+	}
+	
+	/**
+		Deserializes an object from filesystem.
+		@return deserialized object.
+	*/
+	protected Object read(String fileName)	{
+		fileName = makeFilePath(fileName);
+		System.err.println("deserializing object from "+fileName);
+		ObjectInputStream oin = null;
+		try	{
+			FileInputStream in = new FileInputStream(fileName);
+			oin = new ObjectInputStream(in);
+			return oin.readObject();
+		}
+		catch (Exception e)	{
+			System.err.println(e.getMessage());	// tolerate non-existing object
+			return null;
+		}
+		finally	{
+			try	{	oin.close();	}	catch (Exception e)	{}
+		}
+	}
+	
+	/**
+		Serializes an object to filesystem.
+		@return true on success.
+	*/
+	protected boolean write(String fileName, Object o)	{
+		fileName = makeFilePath(fileName);
+		System.err.println("serializing object to "+fileName);
+		ObjectOutputStream oout = null;
+		try	{
+			ensureDirectory(fileName);
+			FileOutputStream out = new FileOutputStream(fileName);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *out*. The resource is referenced at lines 50, 54, 55, 62, 63. The resource is closed at line 63. There are other execution paths that do not contain closure statements, for example, when an exception is thrown by ObjectOutputStream.flush, ObjectOutputStream.writeObject. Ensure *out* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).